src/core/eval.ml

```diff
@@ -138,6 +148,23 @@ and eq_modulo : term -> term -> bool = fun a b ->
     | (Appl(t1,u1), Appl(t2,u2)) -> eq_modulo ((u1,u2)::(t1,t2)::l)
     | (Meta(m1,a1), Meta(m2,a2)) when m1 == m2 ->
         eq_modulo (if a1 == a2 then l else List.add_array2 a1 a2 l)
+    (* Try unfolding variable definitions. *)
+    | ((Vari(x) as t), (Vari(y) as u)) ->
+        if Bindlib.eq_vars x y then eq_modulo l else
+        begin
+          match (def_of x ctx, def_of y ctx) with
+          | None   , None    -> raise Exit
+          | None   , Some(u) -> eq_modulo ((t, u) :: l)
+          | Some(t), None    -> eq_modulo ((t, u) :: l)
+          | Some(t), Some(u) -> eq_modulo ((t, u) :: l)
+        end
```

Comment:
I think this code is useless if you reintroduce the line you removed (equality of variables). Am I wrong?

Reply:
I think so. The previous code didn't handle the case of two distinct variables whose definitions were equal.

Rodolphe, it would be good that you read the comments I made on Gabriel's code first. I asked you to check Gabriel's code because you said in a previous post that you wanted to check it, but I was already on it...

Reply:
> I think so. The previous code didn't handle the case of two distinct variables whose definitions were equal.

I don't know what the previous code was, but I still think I'm right. Assume that  `x` and `y` are distinct, and that the context gives `x := tx` and `y := ty` with `tx` and `ty` convertible. Now since `x` and `y` are not equal, then the case
```ocaml
   | (Vari(x1)   , Vari(x2)   ) when Bindlib.eq_vars x1 x2 -> eq_modulo l
```
does not match. However, the case
```ocaml
  | (Vari(x)    , t          )
```
matches. So we end up making a recursive call to `eq_modulo ((tx, Vari(y)) :: l)`. So if equality cannot be proved directly (e.g., if `tx` is not `Vari(y)`) then we eventually end up matching
```ocaml
  | (t          , Vari(x)    ) ->
```
This leads to a recursive call to `eq_modulo ((tx, ty) :: l)` and since `tx` and `ty` are convertible this is equivalent to `eq_modulo l`.

About our review process: it is in general a good idea for PRs to be reviewed by two persons. Your review lead to many changes and interactions with Gabriel, which I did not all read since they were marked as resolved or outdated. It is perfectly fine to proceed like this, and the code surely improved a lot before I even started my own review. And I think that is the whole point: having a second review (i.e., a fresh pair of eyes on the code) prior to merging is great to catch small errors that could have been overlooked. This is why I wanted to review at this point, and I had the impression that your review was finished given your message and the fact that all comments were resolved.

Reply:
Right. Thank you.

Reply:
It is indeed the old code.